src/compiler/crystal/semantic/call.cr

```diff
@@ -790,18 +790,22 @@ class Crystal::Call
       output_type = program.nil if output_type.void?
     end
 
+    auto_unpack_needed = false
+
     if yield_vars
       # Check if tuple unpacking is needed
-      if yield_vars.size == 1 &&
-         (yield_var_type = yield_vars.first.type).is_a?(TupleInstanceType) &&
-         block.args.size > 1
+      if auto_unpack_needed = (
+           yield_vars.size == 1 &&
+           (yield_var_type = yield_vars.first.type.as?(TupleInstanceType)) &&
+           block.args.size > 1 &&
+           !block.splat_index
+         )
```

Comment:
Stylistically, I would suggest to move such complex assignments out of the condition to a separate expression.